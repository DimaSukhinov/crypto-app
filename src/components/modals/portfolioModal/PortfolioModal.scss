.portfolio {

  &__price {
    border-bottom: 1px solid;
    margin: 5px 0;
  }

  &__values {
    display: flex;
    align-items: center;
  }

  &__element {
    margin-right: 60px;
    padding: 15px;
  }

  &__name, &__price-usd {
    min-width: 130px;
  }

  &__amount {
    min-width: 30px;
  }

  &__delete-button {
    border: 1px solid;
    border-radius: 100%;
    padding: 0 8px;

    &:hover {
      background-color: #edeef0;
      cursor: pointer;
    }
  }

  &__delete-modal {
    display: flex;
    justify-content: space-around;
  }

  &__empty-portfolio {
    font-size: 22px;
    margin: 30px 50px;
  }
}
