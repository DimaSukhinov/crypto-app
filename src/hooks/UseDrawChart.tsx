import {useCallback, useEffect, useState} from 'react';
import {GraphicDataType} from '../components/values/value/Value';
import {cryptoAPI} from '../api/api';

export const useDrawChart = (value: string) => {

    const [data, setData] = useState<GraphicDataType[]>([])
    const [chartData, setChartData] = useState<GraphicDataType[]>([])
    const [chartValue, setChartValue] = useState<'day' | '2days'>('day')

    useEffect(() => {
        cryptoAPI.graphic(value).then((data) => {
            setData(data.data.data)
            setChartData(data.data.data.reverse().slice(0, 24).reverse())
        })
    }, [value])

    const drawChart = useCallback((data: any, value: 'day' | '2days') => () => {
        setChartData(data)
        setChartValue(value)
    }, [])

    return {data, chartValue, chartData, drawChart}
}
