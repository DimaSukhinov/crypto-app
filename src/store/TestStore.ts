import { combineReducers, createStore } from 'redux';
import { ValueType } from './values-reducer';
import { PortfolioType } from './portfolio-reducer';

export const values = [
  {
    id: 'bitcoin', rank: '1', changePercent24Hr: '3.1367184772434482', explorer: 'https://blockchain.info/',
    marketCapUsd: '452445368357.8622563021714334', maxSupply: '21000000.0000000000000000', name: 'Bitcoin',
    priceUsd: '23679.9797033025578357', supply: '19106662.0000000000000000', symbol: 'BTC',
    volumeUsd24Hr: '17521511686.5773940471774064', vwap24Hr: '23710.9068155355389843',
  },
  {
    id: 'ethereum', rank: '2', changePercent24Hr: '3.1367184772434482', explorer: 'https://blockchain.info/',
    marketCapUsd: '452445368357.8622563021714334', maxSupply: '21000000.0000000000000000', name: 'Ethereum',
    priceUsd: '23679.9797033025578357', supply: '19106662.0000000000000000', symbol: 'BTC',
    volumeUsd24Hr: '17521511686.5773940471774064', vwap24Hr: '23710.9068155355389843',
  },
];

export const portfolio: PortfolioType[] = [
  { id: 'bitcoin', name: 'Bitcoin', price: 2000, valueCount: 20 },
  { id: 'ethereum', name: 'Ethereum', price: 2000, valueCount: 20 },
];

const ValuesState: ValueType[] = [
  {
    'id': 'bitcoin',
    'rank': '1',
    'symbol': 'BTC',
    'name': 'Bitcoin',
    'supply': '19109256.0000000000000000',
    'maxSupply': '21000000.0000000000000000',
    'marketCapUsd': '439598186222.6843028478185096',
    'volumeUsd24Hr': '9622382958.4556673270153307',
    'priceUsd': '23004.4637123854692641',
    'changePercent24Hr': '-3.4084952567069450',
    'vwap24Hr': '23579.4890013699165648',
    'explorer': 'https://blockchain.info/',
  },
  {
    'id': 'ethereum',
    'rank': '2',
    'symbol': 'ETH',
    'name': 'Ethereum',
    'supply': '121784637.9365000000000000',
    'maxSupply': null,
    'marketCapUsd': '200976786498.8091803887365518',
    'volumeUsd24Hr': '6534517118.1570715975822179',
    'priceUsd': '1650.2638584318075930',
    'changePercent24Hr': '-3.8175749864566575',
    'vwap24Hr': '1696.1297775541851926',
    'explorer': 'https://etherscan.io/',
  },
  {
    'id': 'tether',
    'rank': '3',
    'symbol': 'USDT',
    'name': 'Tether',
    'supply': '66197977707.2567140000000000',
    'maxSupply': null,
    'marketCapUsd': '66250453592.6509238582810531',
    'volumeUsd24Hr': '20837775336.8158968284416711',
    'priceUsd': '1.0007927113064733',
    'changePercent24Hr': '0.1010078632835212',
    'vwap24Hr': '1.0003521044485800',
    'explorer': 'https://www.omniexplorer.info/asset/31',
  },
  {
    'id': 'usd-coin',
    'rank': '4',
    'symbol': 'USDC',
    'name': 'USD Coin',
    'supply': '54399430698.1528200000000000',
    'maxSupply': null,
    'marketCapUsd': '54492359288.5810234021565036',
    'volumeUsd24Hr': '1442267903.9112855456361948',
    'priceUsd': '1.0017082640247439',
    'changePercent24Hr': '0.1224038365462283',
    'vwap24Hr': '0.9993086427840540',
    'explorer': 'https://etherscan.io/token/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
  },
  {
    'id': 'binance-coin',
    'rank': '5',
    'symbol': 'BNB',
    'name': 'BNB',
    'supply': '166801148.0000000000000000',
    'maxSupply': '166801148.0000000000000000',
    'marketCapUsd': '47284799962.0041650982695676',
    'volumeUsd24Hr': '370241395.7990609671005731',
    'priceUsd': '283.4800631108615937',
    'changePercent24Hr': '-3.4432572668922072',
    'vwap24Hr': '288.5517843856753446',
    'explorer': 'https://etherscan.io/token/0xB8c77482e45F1F44dE1745F52C74426C631bDD52',
  },
];

const PortfolioState: PortfolioType[] = [
  { id: 'bitcoin', name: 'Bitcoin', price: 2000, valueCount: 20 },
  { id: 'ethereum', name: 'Ethereum', price: 2000, valueCount: 20 },
];

const rootReducer = combineReducers({
  values: (state: ValueType[] = ValuesState): ValueType[] => {
    return state;
  },
  portfolio: (state: PortfolioType[] = PortfolioState): PortfolioType[] => {
    return state;
  },
});

export const testStore = createStore(rootReducer);
