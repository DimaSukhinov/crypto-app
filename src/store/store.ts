import {combineReducers, createStore} from 'redux';
import {TypedUseSelectorHook, useSelector} from 'react-redux';
import {valuesReducer} from './values-reducer';
const rootReducer = combineReducers({
    values: valuesReducer,
})

export const store = createStore(rootReducer)

export type AppRootStateType = ReturnType<typeof rootReducer>

export const useAppSelector: TypedUseSelectorHook<AppRootStateType> = useSelector

// @ts-ignore
window.store = store
