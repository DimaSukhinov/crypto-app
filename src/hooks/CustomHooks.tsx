import { TypedUseSelectorHook, useSelector } from 'react-redux';
import { AppRootStateType } from '../store/store';

export const useAppSelector: TypedUseSelectorHook<AppRootStateType> = useSelector;
