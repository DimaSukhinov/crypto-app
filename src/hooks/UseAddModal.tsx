import {ChangeEvent, useCallback} from 'react';
import {addToPortfolioAC, PortfolioType} from '../store/portfolio-reducer';
import {useDispatch} from 'react-redux';

export const useAddModal = (
    setActiveAddModal: (activeAddModal: boolean) => void,
    setCurrentValue: (currentValue: string) => void,
    setError: (error: boolean) => void,
    setValueCount: (valueCount: number) => void
) => {

    const dispatch = useDispatch()

    const openAddModal = useCallback((id: string) => {
        setActiveAddModal(true)
        setCurrentValue(id)
        setError(false)
        setValueCount(0)
    }, [setActiveAddModal, setCurrentValue, setError, setValueCount])

    const onValueCountChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
        setValueCount(+e.currentTarget.value)
        setError(false)
    }, [setError, setValueCount])

    const addToPortfolio = useCallback((id: string, name: string, price: string, valueCount: number) => () => {
        if (valueCount > 0) {
            let value: PortfolioType = {id: id, name: name, price: +price, valueCount: valueCount}
            dispatch(addToPortfolioAC(value))
            setActiveAddModal(false)
            setValueCount(0)
        } else setError(true)
    }, [dispatch, setActiveAddModal, setError, setValueCount])

    const closeModal = useCallback(() => setActiveAddModal(false), [setActiveAddModal])

    return {
        openAddModal, onValueCountChange, addToPortfolio, closeModal
    }
}
