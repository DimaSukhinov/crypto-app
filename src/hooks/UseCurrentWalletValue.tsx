import React from 'react';
import { ValueType } from '../store/values-reducer';
import { PortfolioType } from '../store/portfolio-reducer';

export const useCurrentWalletValue = (portfolio: PortfolioType[], values: ValueType[]) => {

  const currentWalletValueArr = portfolio.map(p => values
    .map(v => v.id === p.id && +(p.valueCount * +v.priceUsd)).reduce((acc: any, num: any) => acc + num, 0));
  const currentWalletValue = currentWalletValueArr.reduce((acc: any, num: any) => acc + num, 0).toFixed(2);
  const originalWalletValue = portfolio.map(p => p !== null && p.valueCount * p.price).reduce((acc: any, num: any) => acc + num, 0).toFixed(2);
  const walletDifference = currentWalletValue - originalWalletValue;
  const walletDifferencePercent = walletDifference / currentWalletValue;

  return <>
    {currentWalletValue + ' USD '}
    {currentWalletValue > 0 && <span style={{ color: walletDifference > 0 ? '#0abd0a' : '#cb2424' }}>
            {walletDifference.toFixed(2)} $ ({walletDifferencePercent.toFixed(4)} %)
        </span>}
  </>;
};
