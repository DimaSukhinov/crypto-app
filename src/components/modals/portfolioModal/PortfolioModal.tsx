import React, {useCallback, useState} from 'react';
import './PortfolioModal.scss';
import {Modal} from '../Modal';
import {ConfirmDeletionModal} from '../confirmDeletionModal/ConfirmDeletionModal';
import {useDispatch} from 'react-redux';
import {removeFromPortfolioAC} from '../../../store/portfolio-reducer';
import {PortfolioValue} from './portfolioValue/PortfolioValue';
import {useAppSelector} from '../../../hooks/CustomHooks';

type PortfolioModalPropsType = {
    currentWalletValue: React.ReactNode
    activePortfolioModal: boolean
    setActivePortfolioModal: (activePortfolioModal: boolean) => void
}

export const PortfolioModal = React.memo(({
                                              activePortfolioModal, setActivePortfolioModal, currentWalletValue
                                          }: PortfolioModalPropsType) => {

    const dispatch = useDispatch()
    const portfolio = useAppSelector((store) => store.portfolio)
    const [valueForDelete, setValueForDelete] = useState<string>('')
    const [confirmDeletionModal, setConfirmDeletionModal] = useState<boolean>(false)

    const removeValueFromPortfolio = useCallback((id: string) => () => {
        setValueForDelete(id)
        setConfirmDeletionModal(true)
    }, [])

    const closePortfolioModal = useCallback(() => setActivePortfolioModal(false), [setActivePortfolioModal])

    const confirmValueDelete = useCallback(() => {
        dispatch(removeFromPortfolioAC(valueForDelete))
        setConfirmDeletionModal(false)
    }, [dispatch, setConfirmDeletionModal, valueForDelete])

    const rejectValueDelete = useCallback(() => setConfirmDeletionModal(false), [])

    return <>
        {activePortfolioModal && <Modal closeModal={closePortfolioModal}>
            {portfolio.length > 0
                ? <div className={'portfolio'}>
                    <div className={'portfolio__price'}>
                        Current price: {currentWalletValue}
                    </div>
                    <PortfolioValue portfolio={portfolio} removeValueFromPortfolio={removeValueFromPortfolio}/>
                    {confirmDeletionModal &&
                        <ConfirmDeletionModal valueForDelete={valueForDelete} confirm={confirmValueDelete}
                                              reject={rejectValueDelete}/>}
                </div>
                : <div className={'portfolio__empty-portfolio'}>You don't have currency</div>}
        </Modal>}
    </>
})
