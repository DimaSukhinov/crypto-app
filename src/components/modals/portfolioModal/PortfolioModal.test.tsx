import React from 'react';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {Provider} from 'react-redux';
import {store} from '../../../store/store';
import {PortfolioModal} from './PortfolioModal';
import {useCurrentWalletValue} from '../../../hooks/UseCurrentWalletValue';
import {portfolio, testStore, values} from '../../../store/TestStore';
import userEvent from '@testing-library/user-event';

describe('Portfolio modal', () => {

    test('empty portfolio', () => {
        const currentWalletValue = useCurrentWalletValue([], [])
        render(
            <Provider store={store}>
                <PortfolioModal setActivePortfolioModal={jest.fn()} activePortfolioModal={true}
                                currentWalletValue={currentWalletValue}/>
            </Provider>
        )

        expect(screen.getByTestId('portfolio-modal')).toBeInTheDocument() // portfolio modal expected
        expect(screen.getByTestId('portfolio-modal')).toMatchSnapshot()
        expect(screen.getByText(/you don't have currency/i)).toBeInTheDocument()
    });

    test('not empty portfolio', () => {
        const currentWalletValue = useCurrentWalletValue(portfolio, values)
        render(
            <Provider store={testStore}>
                <PortfolioModal setActivePortfolioModal={jest.fn()} activePortfolioModal={true}
                                currentWalletValue={currentWalletValue}/>
            </Provider>
        )

        const portfolioElem = screen.getAllByTestId('portfolio-elem')

        expect(screen.getByTestId('portfolio-modal')).toBeInTheDocument() // portfolio modal expected
        expect(screen.getByTestId('portfolio-modal')).toMatchSnapshot()

        expect(screen.queryByText(/you don't have currency/i)).not.toBeInTheDocument()
        expect(screen.getByText(/current price/i)).toBeInTheDocument()
        expect(portfolioElem).toHaveLength(2)
    });

    test('expect confirmation window not open', () => {
        const currentWalletValue = useCurrentWalletValue(portfolio, values)
        render(
            <Provider store={testStore}>
                <PortfolioModal setActivePortfolioModal={jest.fn()} activePortfolioModal={true}
                                currentWalletValue={currentWalletValue}/>
            </Provider>
        )

        expect(screen.getByTestId('portfolio-modal')).toBeInTheDocument() // portfolio modal expected
        expect(screen.getAllByTestId('portfolio-elem')).toHaveLength(2)

        expect(screen.queryByTestId('confirm-deletion-modal')).not.toBeInTheDocument()
    });

    test('expect that the confirmation window is open and closes by clicking on the button', () => {
        const currentWalletValue = useCurrentWalletValue(portfolio, values)
        render(
            <Provider store={testStore}>
                <PortfolioModal setActivePortfolioModal={jest.fn()} activePortfolioModal={true}
                                currentWalletValue={currentWalletValue}/>
            </Provider>
        )

        expect(screen.getByTestId('portfolio-modal')).toBeInTheDocument() // portfolio modal expected
        expect(screen.getAllByTestId('portfolio-elem')).toHaveLength(2) // expected 2 values in the portfolio

        userEvent.click(screen.getAllByText('-')[0]) // clicking on the delete button of the first element

        expect(screen.getByTestId('confirm-deletion-modal')).toBeInTheDocument() // confirm deletion modal expected

        userEvent.click(screen.getByText(/no/i)) // click on the button do not delete

        expect(screen.queryByTestId('confirm-deletion-modal')).not.toBeInTheDocument() // confirm deletion modal no expected
        expect(screen.getAllByTestId('portfolio-elem')).toHaveLength(2) // expected 2 values in the portfolio
    });
})
