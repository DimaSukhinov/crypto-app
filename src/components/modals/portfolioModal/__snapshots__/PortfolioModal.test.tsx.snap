// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio modal empty portfolio 1`] = `
<div
  data-testid="portfolio-modal"
>
  <div
    class="reusableModal"
    data-testid="space-around-modal"
  >
    <div
      class="reusableModal__content"
    >
      <div
        class="portfolio__empty-portfolio"
      >
        You don't have currency
      </div>
    </div>
  </div>
</div>
`;

exports[`Portfolio modal not empty portfolio 1`] = `
<div
  data-testid="portfolio-modal"
>
  <div
    class="reusableModal"
    data-testid="space-around-modal"
  >
    <div
      class="reusableModal__content"
    >
      <div
        class="portfolio"
      >
        <div
          class="portfolio__price"
        >
          Current price: 
          947199.19 USD 
          <span
            style="color: rgb(10, 189, 10);"
          >
            867199.19
             $ (
            0.9155
             %)
          </span>
        </div>
        <div
          class="portfolio__value"
          data-testid="portfolio-elem"
        >
          <div
            class="portfolio__name portfolio__element"
          >
            Bitcoin
          </div>
          <div
            class="portfolio__amount portfolio__element"
          >
            20
          </div>
          <div
            class="portfolio__price-usd portfolio__element"
          >
            40000.00
             $
          </div>
          <div
            class="portfolio__delete-button"
          >
            -
          </div>
        </div>
        <div
          class="portfolio__value"
          data-testid="portfolio-elem"
        >
          <div
            class="portfolio__name portfolio__element"
          >
            Ethereum
          </div>
          <div
            class="portfolio__amount portfolio__element"
          >
            20
          </div>
          <div
            class="portfolio__price-usd portfolio__element"
          >
            40000.00
             $
          </div>
          <div
            class="portfolio__delete-button"
          >
            -
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
